### PR TITLE
ZCS-4663:Update Google Guava library to 23.0

### PR DESCRIPTION
**Pull request by sneha-patil-synacor**
Updating Guava library to 23.0 version
Code changes:
1. MapMaker().makeComputingMap is deprecated. Used CacheBuilder instead.
2. Objects.toStringHelper is deprecated. Used MoreObjects.toStringHelper instead.
3. Objects.firstNonNull is deprecated. Used MoreObjects.firstNonNull instead.
4. NullOutputStream is deprecated. Used ByteStreams.nullOutputStream() instead.
5. closeQuietly(Closeable) is deprecated. Handling the closure of closeables explicitly.
6. com.google.common.collect.Iterators.emptyIterator() is deprecated. Used Collections.emptyIterator() instead.
7. InternetDomainName.fromLenient(String domain) is deprecated. Used from(String) instead.

ant test-all passed.

### CODE REVIEW

File: store/src/java/com/zimbra/cs/imap/MemcachedImapCache.java [hunk 1]
```diff
@@ -106,7 +107,10 @@ public Object serialize(ImapFolder folder) throws ServiceException {
             } catch (Exception e) {
                 throw ServiceException.FAILURE("Failed to serialize ImapFolder", e);
             } finally {
-                Closeables.closeQuietly(oout);
+                try {
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/imap/MemcachedImapCache.java]: can use apache commons-io IOUtils.closeQuietly instead.

Comment (fatkudu) [reply store/src/java/com/zimbra/cs/imap/MemcachedImapCache.java]: Or follow the recommendations in the deprecation statement for Closeables.closeQuietly and use "try with resources".  See http://grepcode.com/file/repo1.maven.org/maven2/com.google.guava/guava/15.0/com/google/common/io/Closeables.java?av=f

**Comment by fatkudu** [comment]
Would be easier to review if you did different commits for each type of change.  Too late now really but for the future...

File: store/src/java-test/com/zimbra/cs/index/AbstractIndexStoreTest.java [hunk 2]
```diff
@@ -580,7 +579,10 @@ public void termEnum() throws Exception {
             checkNextTerm(fields, new Term(LuceneFields.L_CONTACT_DATA, "zimbra.com"));
             checkAtEnd(fields, LuceneFields.L_CONTACT_DATA);
         } finally {
-            Closeables.closeQuietly(fields);
+            try {
```
Comment (fatkudu) [inline store/src/java-test/com/zimbra/cs/index/AbstractIndexStoreTest.java]: use try with resources?

File: store/src/java-test/com/zimbra/cs/index/ZimbraQueryTest.java [hunk 3]
```diff
@@ -144,7 +143,10 @@ public void searchResultMode() throws Exception {
         ZimbraQueryResults result = query.execute();
         Assert.assertTrue(result.hasNext());
         Assert.assertEquals(contact.getId(), result.getNext().getItemId());
-        Closeables.closeQuietly(result);
+        try {
```
Comment (fatkudu) [inline store/src/java-test/com/zimbra/cs/index/ZimbraQueryTest.java]: try with resources with all of these.

File: store/src/java/com/zimbra/cs/dav/carddav/Filter.java [hunk 4]
```diff
@@ -304,7 +303,10 @@ public TextMatch(Element elem, Filter parent) {
                 ZimbraLog.dav.warn("can't get target mailbox", e);
                 return result;
             } finally {
-                Closeables.closeQuietly(zqr);
+                try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/dav/carddav/Filter.java]: Use an inner try with resources?

File: store/src/java/com/zimbra/cs/dav/resource/AddressObject.java [hunk 5]
```diff
@@ -424,7 +423,10 @@ private static Contact getContactByUID(DavContext ctxt, String uid, Account acco
             } catch (Exception e) {
                 ZimbraLog.dav.error("can't search for: uid=%s", uid, e);
             } finally {
-                Closeables.closeQuietly(zqr);
+                try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/dav/resource/AddressObject.java]: try with resources.

File: store/src/java/com/zimbra/cs/dav/resource/Attachment.java [hunk 6]
```diff
@@ -113,7 +112,10 @@ public Attachment(String uri, String owner, List<String> tokens, DavContext ctxt
         } catch (Exception e) {
             ZimbraLog.dav.error("can't search for: attachment="+name, e);
         } finally {
-            Closeables.closeQuietly(zqr);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/dav/resource/Attachment.java]: try with resources

File: store/src/java/com/zimbra/cs/dav/resource/ScheduleInbox.java [hunk 7]
```diff
@@ -155,7 +154,10 @@ public ScheduleInbox(DavContext ctxt, Folder f) throws DavException, ServiceExce
         } catch (Exception e) {
             ZimbraLog.dav.error("can't search: uri="+getUri(), e);
         } finally {
-            Closeables.closeQuietly(zqr);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/dav/resource/ScheduleInbox.java]: try with resources.

File: store/src/java/com/zimbra/cs/dav/resource/SearchWrapper.java [hunk 8]
```diff
@@ -139,7 +138,10 @@ public boolean isCollection() {
         } catch (Exception e) {
             ZimbraLog.dav.error("can't search: uri="+getUri(), e);
         } finally {
-            Closeables.closeQuietly(zqr);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/dav/resource/SearchWrapper.java]: try with resources.

File: store/src/java/com/zimbra/cs/gal/GalSearchControl.java [hunk 9]
```diff
@@ -490,7 +490,10 @@ private boolean doLocalGalAccountSearch(Account galAcct) {
             ZimbraLog.gal.warn("search on GalSync account failed for %s", galAcct.getId(), e);
             return false;
         } finally {
-            Closeables.closeQuietly(zqr);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/gal/GalSearchControl.java]: try with resources.

File: store/src/java/com/zimbra/cs/imap/ImapHandler.java [hunk 10]
```diff
@@ -3543,7 +3542,10 @@ private boolean search(String tag, String command, ImapSearch i4search, boolean
                             modseq = Math.max(modseq, hit.getModifiedSequence());
                     }
                 } finally {
-                    Closeables.closeQuietly(zqr);
+                    try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/imap/ImapHandler.java]: try with resources.

File: store/src/java/com/zimbra/cs/imap/TcpImapHandler.java [hunk 11]
```diff
@@ -252,7 +251,10 @@ public void run() {
                         sleep(10 * Constants.MILLIS_PER_SECOND);
                     } catch (InterruptedException e) {
                     }
-                    Closeables.closeQuietly(output);
+                    try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/imap/TcpImapHandler.java]: try with resources.

File: store/src/java/com/zimbra/cs/index/DBQueryOperation.java [hunk 12]
```diff
@@ -445,7 +444,7 @@ public void addTag(Tag tag, boolean bool) {
 
     @Override
     public void close() {
-        Closeables.closeQuietly(luceneOp);
+        luceneOp.close();
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/DBQueryOperation.java]: This might need some safety code round it.  Explore callers.  Can luceneOp ever be null?

File: store/src/java/com/zimbra/cs/index/IndexEditor.java [hunk 13]
```diff
@@ -188,7 +187,10 @@ public void doQuery(QueryRunner runner, boolean dump, int groupBy)
                 outputStream.println("Query ran in " + (endTime-startTime) + " ms");
                 outputStream.println("Displayed a total of " + totalShown + " Hits");
             } finally {
-                Closeables.closeQuietly(res);
+                try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/IndexEditor.java]: try with resources.

File: store/src/java/com/zimbra/cs/index/LuceneIndex.java [hunk 14]
```diff
@@ -276,7 +284,10 @@ public synchronized void warmup() {
         } catch (IOException e) {
             ZimbraLog.search.warn("Failed to warm up", e);
         } finally {
-            Closeables.closeQuietly(searcher);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/LuceneIndex.java]: try with resources.

File: store/src/java/com/zimbra/cs/index/LuceneIndex.java [hunk 15]
```diff
@@ -960,9 +986,15 @@ public void close() throws IOException {
             if (count.decrementAndGet() == 0) {
                 ZimbraLog.search.debug("Close IndexSearcher");
                 try {
-                    Closeables.closeQuietly(luceneSearcher);
+                    try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/LuceneIndex.java]: Would feel happier here with a replacement like closeQuietly in case vars are null?

File: store/src/java/com/zimbra/cs/index/LuceneIndex.java [hunk 16]
```diff
@@ -1067,20 +1102,29 @@ public BrowseTerm nextElement() {
                     try {
                         termEnumeration.next();
                     } catch (IOException e) {
-                        Closeables.closeQuietly(termEnumeration);
+                        try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/LuceneIndex.java]: try with resources could make a chunk of this code much simpler.

File: store/src/java/com/zimbra/cs/index/LuceneQueryOperation.java [hunk 17]
```diff
@@ -404,7 +406,10 @@ private void runSearch() {
                     luceneQuery, topDocsLen, hits.getTotalHits(), System.currentTimeMillis() - start);
         } catch (IOException e) {
             ZimbraLog.search.error("Failed to search query=%s", luceneQuery, e);
-            Closeables.closeQuietly(searcher);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/LuceneQueryOperation.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/index/LuceneQueryOperation.java [hunk 18]
```diff
@@ -440,7 +445,10 @@ private Query expandLazyMultiPhraseQuery(Query query) throws IOException {
                         }
                     }
                 } finally {
-                    Closeables.closeQuietly(itr);
+                    try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/LuceneQueryOperation.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/index/RemoteQueryOperation.java [hunk 19]
```diff
@@ -118,7 +118,7 @@ public ZimbraHit peekNext() throws ServiceException {
 
     @Override
     public void close() throws IOException {
-        Closeables.closeQuietly(results);
+        results.close();
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/RemoteQueryOperation.java]: Make sure is safe or use alternative closeQuietly

File: store/src/java/com/zimbra/cs/index/TermsFilter.java [hunk 20]
```diff
@@ -85,7 +84,10 @@ public DocIdSet getDocIdSet(IndexReader reader) throws IOException {
                 }
             }
         } finally {
-            Closeables.closeQuietly(td);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/TermsFilter.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/index/ZimbraQuery.java [hunk 21]
```diff
@@ -788,8 +788,11 @@ public ZimbraQueryResults execute() throws ServiceException {
             }
             return results;
         } catch (RuntimeException e) {
-            Closeables.closeQuietly(results);
-            Closeables.closeQuietly(operation);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/ZimbraQuery.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/index/analysis/RFC822AddressTokenStream.java [hunk 22]
```diff
@@ -133,7 +133,7 @@ private void tokenize(String src, Set<String> emails) {
         add(domain);
 
         try {
-            String top = InternetDomainName.fromLenient(domain).topPrivateDomain().parts().get(0);
+            String top = InternetDomainName.from(domain).topPrivateDomain().parts().get(0);
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/index/analysis/RFC822AddressTokenStream.java]: Is this a safe change?  Needs to be in its own commit with justification.

File: store/src/java/com/zimbra/cs/mailbox/ContactAutoComplete.java [hunk 23]
```diff
@@ -857,7 +857,10 @@ private void queryFolders(String str, String generatedQuery, Map<ItemId, Mountpo
                 }
             }
         } finally {
-            Closeables.closeQuietly(qres);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/mailbox/ContactAutoComplete.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java [hunk 24]
```diff
@@ -283,13 +282,19 @@ public boolean existsInContacts(Collection<InternetAddress> addrs) throws IOExce
                             }
                         }
                     } finally {
-                        Closeables.closeQuietly(values);
+                        try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: try with resources or alternative closeQuietly in a number of places in this file.

File: store/src/java/com/zimbra/cs/mailbox/calendar/CalendarMailSender.java [hunk 25]
```diff
@@ -817,7 +816,10 @@ public static MimeMessage createForwardedPrivateInviteMessage(
             } catch (IOException e) {
                 throw ServiceException.FAILURE("Error writing iCalendar", e);
             } finally {
-                Closeables.closeQuietly(writer);
+                try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/mailbox/calendar/CalendarMailSender.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/pop3/Pop3Mailbox.java [hunk 26]
```diff
@@ -95,7 +94,10 @@
                     }
                 }
             } finally {
-                Closeables.closeQuietly(results);
+                try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/pop3/Pop3Mailbox.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/service/formatter/ArchiveFormatter.java [hunk 27]
```diff
@@ -362,12 +361,18 @@ else if (context.getEndTime() - context.getStartTime() > maxDays * Constants.MIL
                             }
                             aos = saveItem(context, results.getNext().getMailItem(), fldrs, cnts, false, aos, encoder, names);
                         }
-                        Closeables.closeQuietly(results);
+                        try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/formatter/ArchiveFormatter.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/service/formatter/Formatter.java [hunk 28]
```diff
@@ -303,7 +302,10 @@ public void remove() {
         }
 
         public void finished() {
-            Closeables.closeQuietly(results);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/formatter/Formatter.java]: Check for safety - is results ever null?

File: store/src/java/com/zimbra/cs/service/mail/ApplyFilterRules.java [hunk 29]
```diff
@@ -124,7 +123,10 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
                 String msg = String.format("Unable to run search for query: '%s'", query);
                 throw ServiceException.INVALID_REQUEST(msg, e);
             } finally {
-                Closeables.closeQuietly(results);
+                try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/mail/ApplyFilterRules.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/service/mail/GetMsg.java [hunk 30]
```diff
@@ -65,6 +65,12 @@ protected boolean checkMountpointProxy(Element request) {
 
     @Override
     public Element handle(Element request, Map<String, Object> context) throws ServiceException {
+        /*URL[] urls1 = ((URLClassLoader) (Thread.currentThread().getContextClassLoader())).getURLs();
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/mail/GetMsg.java]: Why is this here?  commented out code is yuck.

File: store/src/java/com/zimbra/cs/service/mail/GetShareNotifications.java [hunk 31]
```diff
@@ -114,7 +113,10 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
                 }
             }
         } finally {
-            Closeables.closeQuietly(zqr);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/mail/GetShareNotifications.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/service/mail/Search.java [hunk 32]
```diff
@@ -118,7 +117,10 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
             putHits(zsc, octxt, response, results, params, memberOfMap);
             return response;
         } finally {
-            Closeables.closeQuietly(results);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/mail/Search.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/service/mail/SearchConv.java [hunk 33]
```diff
@@ -114,7 +113,10 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
                         // call me AFTER putHits since some of the <info> is generated by the getting of the hits!
                         builder.add(results.getResultInfo());
             } finally {
-                Closeables.closeQuietly(results);
+                try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/mail/SearchConv.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/service/mail/ToXML.java [hunk 34]
```diff
@@ -2849,6 +2850,9 @@ private static void addContent(Element elt, MPartInfo mpi, int maxSize, boolean
                     wasTruncated = tw.wasTruncated();
                 }
                 ByteUtil.closeStream(stream);
+                /*Class klass = com.google.common.io.Closeables.class;
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/mail/ToXML.java]: Why is this here?

File: store/src/java/com/zimbra/cs/session/SoapSession.java [hunk 35]
```diff
@@ -1628,7 +1623,10 @@ private void addDeletedNotification(ModificationKey mkey, StringBuilder deletedI
 
     public synchronized void clearCachedQueryResults() {
         try {
-            Closeables.closeQuietly(queryResults);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/session/SoapSession.java]: alternative closeQuietly

File: store/src/java/com/zimbra/cs/util/SpamExtract.java [hunk 36]
```diff
@@ -413,7 +412,10 @@ private static void extract(String authToken, Account account, Server server, St
                 }
             }
         } finally {
-            Closeables.closeQuietly(tgzStream);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/util/SpamExtract.java]: try with resources or alternative closeQuietly

File: store/src/java/com/zimbra/cs/zookeeper/CuratorManager.java [hunk 37]
```diff
@@ -169,9 +168,12 @@ public synchronized void start() throws Exception {
 
     public synchronized void stop() {
         if (client != null) {
-            Closeables.closeQuietly(serviceCache);
-            Closeables.closeQuietly(serviceDiscovery);
-            Closeables.closeQuietly(client);
+            try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/zookeeper/CuratorManager.java]: alternative closeQuietly

File: store/src/java/com/zimbra/qa/unittest/TestCalDav.java [hunk 38]
```diff
@@ -485,7 +484,10 @@ public String exampleCancelIcal(Account organizer, Account attendee1, Account at
         StringWriter calWriter = new StringWriter();
         vcal.toICalendar(calWriter);
         String icalString = calWriter.toString();
-        Closeables.closeQuietly(calWriter);
+        try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/qa/unittest/TestCalDav.java]: try with resources

File: store/src/java/com/zimbra/qa/unittest/TestJaxb.java [hunk 39]
```diff
@@ -128,7 +129,12 @@ public void tearDown() throws Exception {
     }
 
     private static String getCN(ZMailbox mbox) throws ServiceException {
-        String cn = mbox.getAccountInfo(false).getAttrs().get("cn").get(0);
+        String cn = null;
+        try {
+            cn = mbox.getAccountInfo(false).getAttrs().get("cn").get(0);
+        } catch (ExecutionException e) {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/qa/unittest/TestJaxb.java]: As suggested elsewhere, should use something that handles ExecutionException at a higher level.

File: store/src/java/com/zimbra/qa/unittest/TestUtil.java [hunk 40]
```diff
@@ -461,7 +460,10 @@ public static ZOutgoingMessage getOutgoingMessage(String recipient, String subje
             ZimbraHit hit = r.getNext();
             ids.add(new Integer(hit.getItemId()));
         }
-        Closeables.closeQuietly(r);
+        try {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/qa/unittest/TestUtil.java]: try with resources or alternative closeQuietly in this file

File: store/src/java/com/zimbra/soap/SoapServlet.java [hunk 41]
```diff
@@ -110,11 +112,21 @@ public void init() throws ServletException {
...
         // See if any extra services were previously added by extensions
         synchronized (sExtraServices) {
-            List<DocumentService> services = sExtraServices.get(getServletName());
-            for (DocumentService service : services) {
-                addService(service);
-                i++;
+            List<DocumentService> services;
+            try {
+                services = sExtraServices.get(getServletName());
+                ZimbraLog.mailbox.info("getServletName()::"+getServletName());
```
Comment (fatkudu) [inline store/src/java/com/zimbra/soap/SoapServlet.java]: Please use format style.

File: store/src/java/com/zimbra/soap/SoapServlet.java [hunk 42]
```diff
@@ -110,11 +112,21 @@ public void init() throws ServletException {
...
+                ZimbraLog.mailbox.info("services::"+services);
+                if(services != null)
```
Comment (fatkudu) [inline store/src/java/com/zimbra/soap/SoapServlet.java]: per coding standards, all if statement bodies should be in {}

**Comment by sneha-patil-synacor** [comment]
Closing this PR and opening a new one with separate StringHelper commit
